omc3/scripts/kmod_lumi_imbalance.py

```diff
@@ -0,0 +1,237 @@
+            df, 
+            save_index=NAME
+        )
+    
+    return df
+
+
+def _read_and_check_dataframes(opt: DotDict) -> dict[str, tfs.TfsDataFrame]:
+    """
+    Read the given DataFrames if needed, check them for validity and return a dictionary.
+    """
+    dfs = {}
+    for ip in IPS:
+        df = opt.get(ip, None)
+        if df is None:
+            continue
+
+        if isinstance(df, (Path, str)):
+            try:
+                df = tfs.read(df, index=BEAM)
+            except KeyError as e:
+                msg = (
+                    f"Dataframe '{df}' does not contain a '{BEAM}' column."
+                    "You need to run the `kmod_average` script on data for both beams "
+                    "before you can calulate the luminosity imbalance."
+                )
+                raise KeyError(msg) from e
+
+        if BEAM in df.columns:
+            df = df.set_index(BEAM)
+
+        for beam in (1, 2):
+            if beam not in df.index:
+                msg = (
+                    f"Dataframe '{df}' does not seem to contain data per beam."
+                    "You need to run the `kmod_average` script on data for both beams "
+                    "before you can calulate the luminosity imbalance."
+                )
+                raise KeyError(msg)
+
+        dfs[ip] = df
+    
+    if len(dfs) != 2:
+        msg = (
+            "Lumi inbalance can only be calculated for exactly two IPs, "
+            f"but instead {len(dfs)} were given."
+        )
+        raise ValueError(msg)
+
+    return dfs
+
+
+def get_lumi_imbalance_df(**kwargs) -> tfs.TfsDataFrame:
+    """
+    Calculate the effective beta stars and the luminosity imbalance from the input dataframes.
+
+    Args:
+        ipA (tfs.TfsDataFrame): ar`TfsDataFrame` with the averaged results from a kmod analysis, for IP_A.
+        ipB (tfs.TfsDataFrame): a `TfsDataFrame` with the averaged results from a kmod analysis, for IP_B.
+        (Actually, any name that ends with an integer is fine.)
+    
+    Returns:
+        tfs.TfsDataFrame with effective beta stars per IP and the luminosity imbalance added to the header.
+    """
+    df_effective_betas = tfs.TfsDataFrame()
+    for ip, df in kwargs.items():
+        df_effective_betas.loc[ip.upper(), [f'{BETASTAR}', f'{ERR}{BETASTAR}']] = get_effective_beta_star_w_err(df)
+    
+    ip_a, ip_b = df_effective_betas.index
+    lumi_imb, lumi_imb_err = get_imbalance_w_err(
+        *tuple(df_effective_betas.loc[ip_a, :]), 
+        *tuple(df_effective_betas.loc[ip_b, :])
+    )
+    
+    df_effective_betas.headers[f'{LUMINOSITY}{IMBALACE}'] = lumi_imb
+    df_effective_betas.headers[f'{ERR}{LUMINOSITY}{IMBALACE}'] = lumi_imb_err
+    return df_effective_betas
+
+
+def get_imbalance_w_err(ipA_beta: float, ipA_beta_err: float, ipB_beta: float, ipB_beta_err: float) -> tuple[float, float]:
+    """
+    Calculate the luminosity imbalance IP_A / IP_B  and its error.
+    """
```

Comment:
We are making the assumption that every other parameter in the Lumi calculation is the same between the two IPs here, right? We might want to at least write that in the docstring, because when we go to flat optics / crabbing etc with HighLumi this is going to be wrong.

Reply:
flat optics should be fine, as this is taken into account. With crabbing I am not sure how this affects the beta function. But sure, I can add a comment.